src/client/clientmap.cpp

```diff
@@ -372,9 +372,16 @@ void ClientMap::updateDrawList()
 		}
 
 		// Block meshes are stored in blocks where all coordinates are even (lowest bit set to 0)
-		// We add both the current block and the block with mesh to the set.
-		shortlist.emplace(block_coord.X, block_coord.Y, block_coord.Z);
-		shortlist.emplace(block_coord.X & ~1, block_coord.Y & ~1, block_coord.Z & ~1);
+		// Add them to the de-dup set.
+		v3s16 mesh_coord(block_coord.X & ~1, block_coord.Y & ~1, block_coord.Z & ~1);
+		if (mesh_coord == block_coord) {
+			shortlist.insert(mesh_coord);
+		}
+		// All other blocks we can grab and add to the drawlist right away.
+		else if (block && m_drawlist.emplace(block_coord, block).second) {
+			// only grab the ref if the block exists and was not in the list
+			block->refGrab();
```

Comment:
@x2048 Is this safety check needed? I.e. can we see the same block multiple times?

Reply:
Hmm... It's not quite right. The MapBlockComparer is not unique.